README.md

```diff
@@ -13,175 +13,413 @@ Compartmentalization of host behavior hooks for JS.
+
+Defining a module loader in the language also improves the language's ability
+to evolve.
+For example, a module loader interface that accounts for linking third-party
+modules that are not JavaScript facilitates easier experimentation with linkage
+against languages like WASM.
+For another, a module loader interface allows for user space experimentation
+with the notion of [import maps][import-map].
+
+Defining a module loader in the language also provides valuable insight to the
+design of every language feature that touches upon modules, and every new
+module system feature adds uncertainty to the eventual inclusion of a module
+loader to the language.
+
+One such insight is that module blocks will benefit from the notion of a module
+descriptor as defined by this proposal. Module blocks roughly correspond to
+compiled sources and are consequently not coupled to a particular host environment.
+A module descriptor is necessary to carry properties of a module not captured
+in the source, like the module referrer specifier and how to populate `import.meta`.
+
+Additionally, having a module loader interface is a prerequisite for shimming built-in
+modules.
 
 ### Sketch
 
-This is a rough sketch of potential APIs.
+This is a rough sketch of potential interfaces.
 
 ```ts
-// Shared space by Realm
-type FullSpecifier = string;
-type ModuleNamespace = object;
-// Used to create a Reusable Instance factory for Module Records
-// exotic
-interface StaticModuleRecord {
-  // intend to add reflection of import/export bindings
+interface ModuleExportsNamespace {}
+
+// Module instances have a module environment record, an exotic object
+// that reflects every name in the lexical scope of the module.
+// The environment record does not contain a property for any names that are
+// imported and reexported without a lexical binding.
+// An `import name as alias` binding will have a property with the lexically bound alias.
+// An `export name as alias` binding will have a property with the lexically bound name,
+// whereas the module exports namespace will have a property with the alias.
+interface ModuleEnvironmentRecord {}
+
+// Bindings reflect the `import` and `export` statements of a module.
+type Binding =
+  { import: '*' | string, as?: string, from: string } |
+  { export: '*' | string, as?: string, from?: string };
+
+// Compartments support ECMAScript modules and linkage to other kinds of modules,
+// notably allowing for JSON or WASM.
+// These must provide an initializer function and may have bindings.
+// The bindings correspond to the equivalent `import` and `export` declarations
+// of an ECMAScript module.
+type ThirdPartyStaticModuleRecord = {
+  bindings?: Array<Binding>,
+  // Initializes the module if it is imported.
+  // Initialize may return a promise, indicating that the module uses
+  // the equivalent of top-level-await.
+  initialize(environment: ModuleEnvironmentRecord, importMeta: Object | undefined),
```

Comment:
```suggestion
  initialize(environment: ModuleEnvironmentRecord, importMeta: Object | undefined, dynamicImport: (spec: string) => Promise<ModuleExportsNamespace>),
```

Reply:
@patrick-soquet I will follow-up with a change here that will interest you:

Jack brings up an interesting point that we need to thread `import` and `import.meta`. Consequently, we probably ought to have both `needsImport` and `needsImportMeta` flags on the third-party static-module-record. And since both are optional, the `initialize` function should probably take an options bag.

Reply:
😂 one more step forwards SystemJS format

Reply:
😂 one more step forwards SystemJS format